modules/services/dnscrypt-proxy.nix

```diff
@@ -0,0 +1,56 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+
+  cfg = config.services.dnscrypt-proxy;
+
+  format = pkgs.formats.toml { };
+
+  configFile = format.generate "dnscrypt-proxy.toml" cfg.settings;
+
+in
+
+{
+  options.services.dnscrypt-proxy = {
+
+    enable = lib.mkEnableOption "Whether to enable the dnscrypt-proxy service.";
+
+    package = lib.mkPackageOption pkgs "dnscrypt-proxy2" { };
+
+    settings = lib.mkOption {
+      description = ''
+        Attrset that is converted and passed as TOML config file.
+        For available params, see: <https://github.com/DNSCrypt/dnscrypt-proxy/blob/${pkgs.dnscrypt-proxy2.version}/dnscrypt-proxy/example-dnscrypt-proxy.toml>
+      '';
+      example = lib.literalExpression ''
+        {
+          sources.public-resolvers = {
+            urls = [ "https://download.dnscrypt.info/resolvers-list/v2/public-resolvers.md" ];
+            cache_file = "public-resolvers.md";
+            minisign_key = "RWQf6LRCGA9i53mlYecO4IzT51TGPpvWucNSCh1CBM0QTaLn73Y7GFO3";
+            refresh_delay = 72;
+          };
+        }
+      '';
+      type = format.type;
+      default = { };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    launchd.daemons.dnscrypt-proxy = {
+      script = ''
+        ${lib.getExe' cfg.package "dnscrypt-proxy"} -config ${configFile}
+      '';
+      serviceConfig = {
+        RunAtLoad = true;
+        KeepAlive = true;
+      };
+    };
```

Comment:
This should have its own user assigned and run as that user; there’s no reason for `dnscrypt-proxy` to run as `root`. See e.g. `modules/services/monitoring/prometheus-node-exporter.nix` for an example.

Reply:
I have been update this PRs with changes system user (following prometheus-node-exporter) @emilazy